### PR TITLE
Make the service worker respond with the fastest between the cache and the network

### LINKED ISSUE #48: Request both from the cache and the network and respond with whatever is faster
Like https://serviceworke.rs/cache-then-network.html.

### PR DESCRIPTION
**Pull request by marco-c**
Fixes #48.

### CODE REVIEW

File: wp-sw-cache/lib/service-worker.js [hunk 1]
```diff
@@ -112,19 +112,28 @@
...
-              this.log('[fetch] Cache hit, returning from ServiceWorker cache: ', event.request.url);
-              return response;
-            }
-            this.log('[fetch] Cache miss, retrieving from server: ', event.request.url);
-            return fetch(event.request);
-          })
-          .catch(e => {
-            this.warn('[fetch] error: ', e);
-          })
+        Promise.race([ fromCache, fromNetwork, ])
```
Comment (darkwing) [inline wp-sw-cache/lib/service-worker.js]: Oh sh*t, forgot about `Promise.race` -- nice!

Comment (darkwing) [inline wp-sw-cache/lib/service-worker.js]: nit but can we remove the trailing comma?

**Comment by darkwing** [comment]
Is this going to put more strain on the site/network?

**Comment by marco-c** [comment]
> Is this going to put more strain on the site/network?

It is, so I was thinking of adding an option, disabled by default, to turn this on. What do you think?

(Although users that install this plugin likely want the best performance possible, so they might want this option on by default).

**Comment by marco-c** [comment]
This should be ready now. r? @darkwing

**Comment by staabm** [comment]
I guess it would make sense to disable on cellular or mobile.

Can a remote request be faster then a local cache lookup?

**Comment by marco-c** [comment]
> Can a remote request be faster then a local cache lookup?

Yes, if you have a slow disk and a fast network.

Maybe we could add heuristics in the future. For example, instead of always racing, we could race for a subset of requests, increasing the subset if the network is faster, decreasing it if the cache is faster.
Or, we could disable racing if the cache is faster than the network for X consecutive requests (where X is a large number).

**Comment by darkwing** [comment]
Rebase :(

**Comment by marco-c** [comment]
Rebased!

**Comment by marco-c** [comment]
@darkwing I'll rebase if you promise to merge.

**Comment by darkwing** [comment]
I promise ... :)

File: wp-offline-shell/wp-offline-shell-admin.php [hunk 2]
```diff
@@ -170,6 +173,13 @@ function options() {
         <input type="checkbox" name="offline_shell_debug" id="offline_shell_debug" value="1" <?php if(intval(get_option('offline_shell_debug'))) echo 'checked'; ?> />
       </td>
     </tr>
+    <tr>
+      <th scope="row"><label for="offline_shell_race_enabled"><?php _e('Enable cache-network race', 'service-worker-cache'); ?></label></th>
+      <td>
+        <input type="checkbox" name="offline_shell_race_enabled" id="offline_shell_race_enabled" value="1" <?php if(intval(get_option('offline_shell_race_enabled'))) echo 'checked'; ?> />
+        <p class="description">Enable this option if you want the service worker to retrieve a response from the cache and the network at the same time, instead of only from the cache. This improves performance for users with fast connections, at the expense of an increased load on your server.</p>
```
Comment (darkwing) [inline wp-offline-shell/wp-offline-shell-admin.php]: Need `__()`

Comment (marco-c) [inline wp-offline-shell/wp-offline-shell-admin.php]: Fixed in fa0fafa.

**Comment by darkwing** [comment]
PR works awesome!  The last thing I can think of that the debug is now confusing:

doubleload

Is there something we can do to only show the debug of whoever won the race?  Just thinking out loud.

**Comment by marco-c** [comment]
4259c45 makes it less confusing.

**Comment by darkwing** [comment]
A few nits but I'll merge and we can address them later.

File: wp-offline-shell/wp-offline-shell-admin.php [hunk 3]
```diff
@@ -185,6 +188,13 @@ function options() {
         <input type="checkbox" name="offline_shell_debug" id="offline_shell_debug" value="1" <?php if(intval(get_option('offline_shell_debug'))) echo 'checked'; ?> />
       </td>
     </tr>
+    <tr>
+      <th scope="row"><label for="offline_shell_race_enabled"><?php _e('Enable cache-network race', 'service-worker-cache'); ?></label></th>
+      <td>
+        <input type="checkbox" name="offline_shell_race_enabled" id="offline_shell_race_enabled" value="1" <?php if(intval(get_option('offline_shell_race_enabled'))) echo 'checked'; ?> />
+        <p class="description"><?php _e('Enable this option if you want the service worker to retrieve a response from the cache and the network at the same time, instead of only from the cache. This improves performance for users with fast connections, at the expense of an increased load on your server.'); ?></p>
```
Comment (marco-c) [inline wp-offline-shell/wp-offline-shell-admin.php]: I forgot to add the text-domain here.

Comment (marco-c) [inline wp-offline-shell/wp-offline-shell-admin.php]: I will add it in the other PR.

**Pull request merged**